src/elfpy/markets.py

```diff
@@ -377,23 +470,148 @@ def _close_long(self, trade_details):
+                    # update their marginal share of cumulative liquidity, give them credit for it (contribution)
+                    contribution[update_agent] += cum_liq_by_agent[update_agent] / cum_liq
+                    # update their owed fees
+                    share_owed[update_agent] = (share_owed[update_agent]*prev_time + contribution[update_agent]*delta_share_fees*delta_time)/current_time
+                    token_owed[update_agent] = (token_owed[update_agent]*prev_time + contribution[update_agent]*delta_token_fees*delta_time)/current_time
+            # update values used for next iteration
+            cum_liq += new_liq
+            cum_liq_by_agent[acting_agent] += new_liq
+            prev_time = current_time
+            prev_share_fees, prev_token_fees = share_fees_till_now, token_fees_till_now
+        if self.verbose:
+            print(f"cum_liq_by_agent = {cum_liq_by_agent}")
+            print(f"      share_owed = {share_owed} calculated sum = {sum(share_owed.values())}"
+                + f"\n                                           vs. direct = {self.share_fees} it's a "
+                + f"match 😁" if sum(share_owed.values()) == self.share_fees else f"mismatch 😢")
+            print(f"      token_owed = {token_owed} calculated sum = {sum(token_owed.values())}"
+                + f"\n                                           vs. direct = {self.token_fees} it's a "
+                + f"match 😁" if sum(token_owed.values()) == self.token_fees else f"mismatch 😢")
+        dur = time.time() - start_time
+        print(f"calc_fees took ", end=f"{fmt(dur,precision=2)} seconds"\
+            if dur>1 else f"{fmt(dur*1000,precision=2)} milliseconds\n")
+        if return_agent is not None:
+            return share_owed[return_agent], token_owed[return_agent]
+
+    def get_market_step_string(self):
+        """Returns a string that describes the current market step"""
+        output_string = f"t={bcolors.HEADER}{self.time}{bcolors.ENDC}"
+        output_string += f" reserves=["\
+                + f"x:{bcolors.OKBLUE}{self.share_reserves*self.share_price}{bcolors.ENDC}"\
+                + f",y:{bcolors.OKBLUE}{self.bond_reserves}{bcolors.ENDC}"\
+                + f",z:{bcolors.OKBLUE}{self.share_reserves}{bcolors.ENDC}"\
+                + f",z_b:{bcolors.OKBLUE}{self.share_buffer}{bcolors.ENDC}"\
+                + f",y_b:{bcolors.OKBLUE}{self.bond_buffer}{bcolors.ENDC}"\
+            + f"]"
+        output_string += f"\n liquidity_pool_history={self.liquidity_pool_history}"
+        output_string += f"\n share_fee_history={self.share_fee_history}"
+        output_string += f"\n token_fee_history={self.token_fee_history}"
+        if self.verbose:
+            output_string += f" fees=["\
+                + f"x:{bcolors.OKBLUE}{self.share_fees}{bcolors.ENDC}"\
+                + f",y:{bcolors.OKBLUE}{self.token_fees}{bcolors.ENDC}"
+        output_string += f" p={bcolors.FAIL}{self.spot_price}{bcolors.ENDC}"
```

Comment:
why do you call this `FAIL`? is the spot price incorrect here?

Reply:
no it's just the color red

Reply:
they're not very descriptive. one is .WARNING (orange, i think)